{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MA770 Mathematical and Statistical Methods of Bioinformatics\n",
    "# Problem Set 6\n",
    "### Cheng, Wanli U31865818"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aperiodic Markov Process\n",
    "### Problem 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For simplicity, we just consider the ordinary Markov process without hidden variables (though we are free add these structure to make it a hidden Markov process).\n",
    "\n",
    "Then, we can consider a simple random walk on a graph. A graph $\\Gamma(V,E)$ is consists of vertices $v_i\\in V$ and edges $e_{ij}\\in E$  connecting $v_i$ and $v_j$. By identifying $e_{ij}$ and $e_{ji}$, we call it a bidirected graph. A simple random walk on a (bidirected) graph is a process such that we start from a vertex than have equal probability to move to any adjacent (i.e., connected by an edge) vertex. Define the states of this process to be the position (i.e, the vertex which we ends up) of each step. Since it is a random process and the probability distribution of each step does not depend on the history, it is a Markov process by definition.\n",
    "\n",
    "Then we consider the $\\mathbb{Z}^2$ as a graph such that each point in $\\mathbb{Z}^2$ is a vertex and define any pair of vertices with Euclidean distance 1 to be adjacent. Now consider the simple random walk on $\\Gamma_{\\mathbb{Z}^2}$. Since there are infinite number of states and the probability of moving to an arbitrary point from a starting point is strictly positive. Therefore, it is an aperiodic Markov Process. Moreover, any simple random walk on an infinite graph can be considered as an aperiodic Markov Process with respect to the position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsimony Conservation Score\n",
    "### Problem 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let $s(\\varphi)$ denote the total mutations needed to reach the observed information of current species if the original information is $\\varphi\\in\\{A,C,G,T\\}$\n",
    "First letter: The result is $(A,A,A,A,A,A)$. We have \n",
    "$$\n",
    "s(A)=0;s(C)=s(G)=s(T)=2.\n",
    "$$\n",
    "Therefore, the parsimony score is $s=0$.\n",
    "\n",
    "Second letter: The result is $(C,C,C,A,A,A)$. We have \n",
    "$$\n",
    "s(A)=s(C)=1;s(G)=s(T)=2.\n",
    "$$\n",
    "Therefore, the parsimony score is $s=1$.\n",
    "\n",
    "Third letter: The result is $(T,G,T,G,T,G)$. We have \n",
    "$$\n",
    "s(A)=s(C)=4;s(G)=s(T)=3.\n",
    "$$\n",
    "Therefore, the parsimony score is $s=3$.\n",
    "\n",
    "Forth letter: The result is $(A,C,G,T,A,A)$. We have \n",
    "$$\n",
    "s(A)=3;s(C)=s(G)=s(T)=5.\n",
    "$$\n",
    "Therefore, the parsimony score is $s=3$.\n",
    "\n",
    "The results in this sections are all correct."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Uniqueness of Phylogentic Tree\n",
    "### Problem 3\n",
    "<b> Theorem: </b>: Given a set $S$ of species and an ultrametric $d$ on $S$, there is a unique tree\n",
    "structure whose leaves are $S$ which is consistent with $d$.\n",
    "\n",
    "Verify case (1) i.e, $d(x_L,x_{k+1})=d(x_R,x_{k+1})$ induce the unique structure of $T_{K+1}$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b> Proof: </b>\n",
    "By hypothesis, $d$ is ultrametric, therefore, we must have $d(x_L,x_{k+1})=d(x_R,x_{k+1})>d(x_L,x_R)$.\n",
    "\n",
    "By construction, $x_L$ and $x_R$ are selected from the left branch and the right branch of the tree $T_k$, which implies that \n",
    "$$\n",
    "d(x_L,x_R)=2d(x_i,r_k)\n",
    "$$\n",
    "for $x_i$ an arbitrary element of $S_k$ and $r_k$ the root of $T_k$.\n",
    "\n",
    "Then we claim that $x_{k+1}$ cannot be an offspring of $r_k$.\n",
    "\n",
    "By Reductio ad Absurdum, assume $x_{k+1}$ to be an offspring of $r_k$. Then by construction, we have \n",
    "$$\n",
    "d(x_{k+1},r_k)=d(x_i,r_k)\n",
    "$$\n",
    "for all $x_i\\in S_k$.\n",
    "\n",
    "Therefore, we have\n",
    "$$\n",
    "d(x_{k+1},x_L)\\le d(x_{k+1},r_k)+d(x_{L},r_k)=2d(x_i,r_k)=d(x_L,x_R),\n",
    "$$\n",
    "which is contradict to our hypothesis.\n",
    "\n",
    "Hence, $x_{k+1}$ is not an offspring of $r_k$. In our context such that there is a connected tree structure $T_{k+1}$ associate to $S_{k+1}$, this fact implies that there is a node $r_{k+1}$ that $r_k$ and $x_{k+1}$ are children of it.\n",
    "\n",
    "The trivial case of $n=2$ tells us the structure $R_{k+1}=\\{\\{r_{k+1},r_k,x_{k+1}\\},\\{\\overrightarrow{r_{k+1}r_k},\\overrightarrow{r_{k+1}x_{k+1}}\\}\\}$ is unique and irreducible. Since $T_k$ by definition, is the unique tree structure on $S_k$, we must have $T_{k+1}=T_k\\cup R_{k+1}$ is unique.\n",
    "\n",
    "By construction $T_{k+1}$ is a rooted (binary) tree, and $S_{k+1}\\subset T_{k+1}$, thus it is the unique tree associate to $S_{k+1}$ and the given ultrametric $d$, which verified our claim."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Continuous time version of the Kimura model\n",
    "### Problem 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Kimura model is given by:\n",
    "$$\n",
    "M=\n",
    "\\begin{bmatrix}\n",
    "1-\\alpha-2\\beta & \\alpha & \\beta & \\beta\\\\\n",
    "\\alpha & 1-\\alpha-2\\beta & \\beta & \\beta\\\\\n",
    "\\beta & \\beta & 1-\\alpha-2\\beta & \\alpha\\\\\n",
    "\\beta & \\beta & \\alpha & 1-\\alpha-2\\beta\\\\\n",
    "\\end{bmatrix}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### （a）\n",
    "The differential form for the Kimura model of evolution can be written as:\n",
    "$$\n",
    "\\frac{d}{dt}\\mathbf{p}(t)=M\\mathbf{p}(t)\n",
    "$$\n",
    "for $\\vec{\\psi}(t)$ denote the probability vector:\n",
    "$$\n",
    "\\mathbf{p}(t)=\n",
    "\\begin{bmatrix}\n",
    "p_A(t)\\\\\n",
    "p_C(t)\\\\\n",
    "p_G(t)\\\\\n",
    "p_T(t)\\\\\n",
    "\\end{bmatrix}\n",
    "=\n",
    "\\begin{bmatrix}\n",
    "P(\\psi(t)=A)\\\\\n",
    "P(\\psi(t)=C)\\\\\n",
    "P(\\psi(t)=G)\\\\\n",
    "P(\\psi(t)=T)\n",
    "\\end{bmatrix}\n",
    "$$\n",
    "such that $\\psi(t)$ denote the explicit state of the gene we are considering at time $t$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)\n",
    "Consider each component of the probability vector, we have:\n",
    "\\begin{equation}\n",
    "\\begin{cases}\n",
    "\\frac{d}{dt}p_A(t)=(1-\\alpha-2\\beta)p_A(t)+\\alpha p_C(t)+\\beta p_G(t)+\\beta p_T(t)\\\\\n",
    "\\frac{d}{dt}p_C(t)=\\alpha p_A(t)+(1-\\alpha-2\\beta)p_C(t)+\\beta p_G(t)+\\beta p_T(t)\\\\\n",
    "\\frac{d}{dt}p_G(t)=\\beta p_A(t)+\\beta p_C(t)+(1-\\alpha-2\\beta) p_G(t)+\\alpha p_T(t)\\\\\n",
    "\\frac{d}{dt}p_T(t)=\\beta p_A(t)+\\beta p_C(t)+\\alpha p_G(t)+(1-\\alpha-2\\beta) p_T(t)\n",
    "\\end{cases}\n",
    "\\end{equation}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (c)\n",
    "By standard matrix methods, we can formulate the continuous time model as:\n",
    "\n",
    "$$\n",
    "\\vec{\\psi}(t)=e^{Mt}\\mathbf{p}(0)\n",
    "$$\n",
    "\n",
    "where\n",
    "\n",
    "$$\n",
    "e^{Mt}=\\sum_{k=0}^{\\infty}\\frac{M^{k}t^{k}}{k!}\n",
    "$$\n",
    "\n",
    "and $\\mathbf{p}(0)$ the probability distribution of the initial state which is given by a random vector $\\mathbf{p}_0$.\n",
    "\n",
    "\n",
    "The spectral decomposition of discrete time Kimura model gives us the following relation:\n",
    "\n",
    "\\begin{equation}\n",
    "M^k=\n",
    "\\frac{1}{4}\n",
    "\\begin{bmatrix}\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\n",
    "\\end{bmatrix}\n",
    "+\\frac{(1-4\\beta)^k}{4}\n",
    "\\begin{bmatrix}\n",
    "1&1&-1&-1\\\\\n",
    "1&1&-1&-1\\\\\n",
    "-1&-1&1&1\\\\\n",
    "-1&-1&1&1\n",
    "\\end{bmatrix}\n",
    "+\\frac{(1-2(\\alpha+\\beta))^k}{2}\n",
    "\\begin{bmatrix}\n",
    "1&-1&0&0\\\\\n",
    "-1&1&0&0\\\\\n",
    "0&0&1&-1\\\\\n",
    "0&0&-1&1\n",
    "\\end{bmatrix}\n",
    "\\end{equation}\n",
    "\n",
    "\n",
    "Therefore, we have:\n",
    "\n",
    "\\begin{align}\n",
    "e^{Mt}\n",
    "&=\\sum_{k=0}^{\\infty}\\frac{M^k}{k!}\\\\\n",
    "&=\n",
    "\\sum_{k=0}^{\\infty}\\left(\n",
    "\\frac{t^k}{4k!}\n",
    "\\begin{bmatrix}\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\n",
    "\\end{bmatrix}\n",
    "+\n",
    "\\frac{(1-4\\beta)^{k}t^{k}}{4k!}\n",
    "\\begin{bmatrix}\n",
    "1&1&-1&-1\\\\\n",
    "1&1&-1&-1\\\\\n",
    "-1&-1&1&1\\\\\n",
    "-1&-1&1&1\n",
    "\\end{bmatrix}\n",
    "+\n",
    "\\frac{(1-2(\\alpha+\\beta))^{k}t^{k}}{2k!}\n",
    "\\begin{bmatrix}\n",
    "1&-1&0&0\\\\\n",
    "-1&1&0&0\\\\\n",
    "0&0&1&-1\\\\\n",
    "0&0&-1&1\n",
    "\\end{bmatrix}\\ \\right)\n",
    "\\\\\n",
    "&=\n",
    "\\sum_{k=0}^{\\infty}\n",
    "\\frac{t^k}{4k!}\n",
    "\\begin{bmatrix}\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\n",
    "\\end{bmatrix}\n",
    "+\n",
    "\\sum_{k=0}^{\\infty}\n",
    "\\frac{(1-4\\beta)^{k}t^{k}}{4k!}\n",
    "\\begin{bmatrix}\n",
    "1&1&-1&-1\\\\\n",
    "1&1&-1&-1\\\\\n",
    "-1&-1&1&1\\\\\n",
    "-1&-1&1&1\n",
    "\\end{bmatrix}\n",
    "+\n",
    "\\sum_{k=0}^{\\infty}\n",
    "\\frac{(1-2(\\alpha+\\beta))^{k}t^{k}}{2k!}\n",
    "\\begin{bmatrix}\n",
    "1&-1&0&0\\\\\n",
    "-1&1&0&0\\\\\n",
    "0&0&1&-1\\\\\n",
    "0&0&-1&1\n",
    "\\end{bmatrix}\n",
    "\\\\\n",
    "&=\n",
    "\\frac{e^t}{4}\n",
    "\\begin{bmatrix}\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\\\\\n",
    "1&1&1&1\n",
    "\\end{bmatrix}\n",
    "+\n",
    "\\frac{e^{(1-4\\beta)t}}{4}\n",
    "\\begin{bmatrix}\n",
    "1&1&-1&-1\\\\\n",
    "1&1&-1&-1\\\\\n",
    "-1&-1&1&1\\\\\n",
    "-1&-1&1&1\n",
    "\\end{bmatrix}\n",
    "+\n",
    "\\frac{e^{(1-2(\\alpha+\\beta))t}}{2}\n",
    "\\begin{bmatrix}\n",
    "1&-1&0&0\\\\\n",
    "-1&1&0&0\\\\\n",
    "0&0&1&-1\\\\\n",
    "0&0&-1&1\n",
    "\\end{bmatrix}\n",
    "\\end{align}\n",
    "\n",
    "Hence, the continuous time version of Kimura model is given by:\n",
    "\n",
    "$$\n",
    "\\mathbf{p}(t)=\n",
    "e^t\n",
    "\\begin{bmatrix}\n",
    "1/4&1/4&1/4&1/4\\\\\n",
    "1/4&1/4&1/4&1/4\\\\\n",
    "1/4&1/4&1/4&1/4\\\\\n",
    "1/4&1/4&1/4&1/4\n",
    "\\end{bmatrix}\n",
    "\\mathbf{p}_0\n",
    "+\n",
    "e^{(1-4\\beta)t}\n",
    "\\begin{bmatrix}\n",
    "1/4&1/4&-1/4&-1/4\\\\\n",
    "1/4&1/4&-1/4&-1/4\\\\\n",
    "-1/4&-1/4&1/4&1/4\\\\\n",
    "-1/4&-1/4&1/4&1/4\n",
    "\\end{bmatrix}\n",
    "\\mathbf{p}_0\n",
    "+\n",
    "e^{(1-2(\\alpha+\\beta))t}\n",
    "\\begin{bmatrix}\n",
    "1/2&-1/2&0&0\\\\\n",
    "-1/2&1/2&0&0\\\\\n",
    "0&0&1/2&-1/2\\\\\n",
    "0&0&-1/2&1/2\n",
    "\\end{bmatrix}\n",
    "\\mathbf{p}_0.\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saitou and Nei algorithm\n",
    "### Problem 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Definition(Phylogentic Distance):</b> \n",
    "Assume there exists a phylogentic tree $T$ define on the set of current species $S=\\{s_1,\\dots,x_n\\}$ and $d$ denote the tree derived distance on $T$.\n",
    "\n",
    "Then define a new distance function on $T$:\n",
    "$$\n",
    "\\delta(x_i,x_j)=(n-4)d(x_i,x_j)-\\sum_{z\\ne x,y}(d(x_i,x_k)+d(x_j,x_k))\n",
    "$$\n",
    "for any $x_i,x_j,x_k\\in T$.\n",
    "\n",
    "<b>Theorem:</b>\n",
    "$x_i,x_j$ are childeren of a single node if and only if\n",
    "$$\n",
    "\\delta(x_i,x_j)=\\delta_{\\min}=\\min\\{\\delta(s_i,s_j)|s_i,s_j\\in S\\}.\n",
    "$$\n",
    "\n",
    "<b>Saito-Nei Algorithm:</b>\n",
    "\n",
    "Given $S=\\{s_1,\\dots,x_n\\}$. Then, assume we have all the information about the tree derived dstance $d(x_i,x_j)$ between any two species $x_i,x_j\\in S$.\n",
    "\n",
    "Then by definition of $\\delta$, we can compute the new distance $\\delta(x_i,x_j)$ for all $x_i,x_j\\in S$ with the given information.\n",
    "\n",
    "Let $T'$ denote some collection of subtrees supportedon $S$, i.e., for any element in $V(T')$, all of its children is contained in $V(T')$ and $S\\subseteq V(T')$. In addition, we define the collection of roots of these subtrees to be $R(T')$.\n",
    "\n",
    "<b>Corollary:</b>\n",
    "For some collection $T'$ of subtrees of $T$, $r_i,r_j\\in R(T')$ are childeren of a single node if and only if\n",
    "$$\n",
    "\\delta(r_i,r_j)=\\delta_\\min^{T'}=\\min\\{\\delta(s_i,s_j)|s_i,s_j\\in R(T')\\}.\n",
    "$$\n",
    "\n",
    "<b>Proof:</b>\n",
    "Let $r_x,r_y,r_z\\in R(T')$ be roots of two trees, and $x,y,z\\in S$ be offsprings of them respectively. Then by construction we can determine the values of $d{x,r_1}$,$d{y,r_2}$ and $d{z,r_3}$ inductively, i,e,\n",
    "$$\n",
    "d(x_i,r_ij)=\\frac{d(x_i,x_k)-d(x_j,x_k)+d(x_i,x_j)}{2}\n",
    "$$\n",
    "for $x_i,x_j$ the immediate children of $r_ij$ and $x_k$ any other species in $S$.\n",
    "\n",
    "Then we have:\n",
    "\\begin{align}\n",
    "d(r_x,r_y)=d(x,y)-d(x,r_x)-d(y,r_y)\\\\\n",
    "d(r_y,r_z)=d(y,z)-d(y,r_y)-d(z,r_z)\\\\\n",
    "d(r_z,r_x)=d(z,x)-d(z,r_z)-d(x,r_x)\n",
    "\\end{align}\n",
    "hence we can unambiguously define $\\delta(r_x,r_y)$.\n",
    "Then we can view the $R(T')$ as a set of current species at some moment in history by making a cutoff at some point. Then the theorem implies that minimizing $\\delta(r_x,r_y)$ coincide with the neighborhood between $r_x$ and $r_y$, which verified our claim.\n",
    "\n",
    "Then, for simplicity, we define\n",
    "$$\n",
    "N(T')=\\{r_i,r_j\\in R(T')|\\delta(r_i,r_j)=\\delta_\\min^{T'}\\}\n",
    "$$\n",
    "to be the set of vertices with neighbours in $T'$.\n",
    "\n",
    "By corollary, we can unambiguously find the node that have pairs in $N(T')$ as children. Hence we can construct a collection of these nodes, $R^{+}(T')$ and the associated edges $E^+(T')$ connected these nodes with their children.\n",
    "\n",
    "<b>Definition:</b> \n",
    "Let the $n^{th}$ *reconstructed tree* to be the following subset of the phylogentic tree $T$:\n",
    "$$\n",
    "T^{n}=T^{n-1}\\cup\\{R^{+}(T^{n-1}), E^+(T^{n-1})\\}\n",
    "$$\n",
    "s.t.\n",
    "$$\n",
    "R(T^{n})=N^{c}(T^{n-1})\\cup R^{+}(T^{n-1})\n",
    "$$\n",
    "where $0^{th}$ *reconstructed tree* is $T_n^0=S$ and $N^c(T^{n-1})=R(T')-N(T')$.\n",
    "\n",
    "The above definition is consistent only if $T^{n}$ is a collection of subtrees of $T$ and containing $S$ for all $n$. We prove it my induction:\n",
    "\n",
    "*Base Step*: It is obvious that $T_n^0=S$ is a collection of subtrees of $T$ and containing $S$.\n",
    "\n",
    "*Assumption*: Assume these condition to be true for $T^n$. \n",
    "\n",
    "*Inductive Step*: By construction $T^{n+1}=T^{n}\\cup\\{R^{+}(T^{n}), E^+(T^{n})\\}$. First, $S\\subseteq T^n$ implies $S\\subseteq T^{n+1}$. Second, by construction, the children of $R^{+}(T^{n})$ are $N(T^{n})\\subseteq R(T^{n}) \\subseteq T^{n}$.\n",
    "\n",
    "Therefore by axiom of mathematical induction, we proved that our definition of $T^n$ is consistent.\n",
    "\n",
    "By construction, $R(T^{n+1})<R(T^n)$, therefore, in a finite step we would reach the root of $T$ from $T^{0}=S$ and reconstruct the whole phylogentic tree, which indicated that the algorithm is effective."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model of Evolution in Fitness space\n",
    "### Problem 6\n",
    "\n",
    "The model of evolution is given by:\n",
    "$$\n",
    "\\frac{d\\mathbf{x}}{dt}=-k\\nabla F(\\mathbf{x})+\\sigma \\mathbf{W}(t)\n",
    "$$\n",
    "where $F$ is the fitness function, \\mathbf{W(t)}=(W_1(t),dots,W_p(t)) is a standard white noise vector, and $k$, $\\sigma$ are constants. In particular, $W_i(t)$ is *standard white noise*, which is given by:\n",
    "$$\n",
    "W_i(t)=\\frac{d}{dt}B(t)\n",
    "$$\n",
    "for $B(t)$ is standard Brownian motion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a)\n",
    "When $\\sigma$ is very small, the model can be approximately considered as the degenerated case\n",
    "$$\n",
    "\\frac{d\\mathbf{x}}{dt}=-k\\nabla F(\\mathbf{x}).\n",
    "$$\n",
    "\n",
    "By definition, the evolution of $\\mathbf{x}$ may converge to a stationary point such that $\\frac{d\\mathbf{x}}{dt}=0$, which implies the following condition:\n",
    "\n",
    "$$\n",
    "-\\nabla F(\\mathbf{x})=0\n",
    "$$\n",
    "\n",
    "i.e., $F$ also reaches its stationary point.\n",
    "\n",
    "Howefver, a stationary point of a function is not necessary the global maximum/minimum of the function. Assume the constant $k$ to be positive, we can only conclude that the point $\\mathbf{x}$ reaches is a local maximum of the fitness function. Therefore, when the local maximum is not the global maximum, $F(\\mathbf{x})$ will not reach the optimal value in this circumstance."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)\n",
    "The possibility of escape a local maximum is depend on the term $\\sigma \\mathbf{W}(t)$.\n",
    "\n",
    "First, to analysis the behavior of $\\sigma \\mathbf{W}(t)$ we assume $k=0$, then\n",
    "$$\n",
    "\\frac{d\\mathbf{x}}{dt}=\\sigma \\mathbf{W}(t),\n",
    "$$\n",
    "which can be written in the component form:\n",
    "$$\n",
    "\\frac{d x_i}{dt}=\\sigma W_i(t)=\\sigma\\frac{d B(t)}{dt},\n",
    "$$\n",
    "since\n",
    "$$\n",
    "\\frac{d\\mathbf{x}}{dt}=(\\frac{\\partial \\mathbf{x}}{\\partial x_i}\\frac{d x_i}{dt})=(\\frac{d x_i}{dt}).\n",
    "$$\n",
    "\n",
    "Solve the differential equation:\n",
    "$$\n",
    "\\frac{d x_i(t)}{dt}=\\sigma\\frac{d B(t)}{dt},\n",
    "$$\n",
    "we have\n",
    "$$\n",
    "x_i(t)=\\sigma B(t)+C.\n",
    "$$\n",
    "Since $x_i(0)=0$ and $B(t)=0$, we have $C=0$. Then the position $x_i(t)=B\\sigma(t)$ follows the distribution:\n",
    "$$\n",
    "x_i(t)\\sim\\mathcal{N}(0,\\sigma^2 t)\n",
    "$$\n",
    "since for Brownian motion, we have\n",
    "$$\n",
    "B(t)-B(s)\\sim\\mathcal{N}(0,t-s).\n",
    "$$\n",
    "\n",
    "Following the suggestion, we assume\n",
    "\\begin{align}\n",
    "F(x)&=-ax+b\\\\\n",
    "-\\frac{d}{dx}F(x)&=a\n",
    "\\end{align}\n",
    "for the one dimensional csae, which implies that\n",
    "\\begin{align}\n",
    "\\frac{d x(t)}{dt}&=ka+\\sigma\\frac{d B(t)}{dt}\\\\\n",
    "x(t)&=kat+\\sigma B(t)+C.\n",
    "\\end{align}\n",
    "With the initial condition, we have $C=0$, hence,\n",
    "\\begin{align}\n",
    "x(t)=kat+\\sigma B(t).\n",
    "\\end{align}\n",
    "Then we have:\n",
    "$$\n",
    "x(t)\\sim\\mathcal{N}(kat,\\sigma^2 t)\n",
    "$$\n",
    "which is still a normal distribution over the whole real line, i.e., $\\forall a\\in\\mathbb{R}, Prob(x(t)=a)>0$. In another words, there is no steepness that can prevent $x$ escaping away.\n",
    "\n",
    "Since the movement in $\\mathbb{R}^3$ can be reduced to three one dimensional movements on the three directions of some orthogonal basis of $\\mathbb{R}^3$, the one dimensional case can be generalize to the 3D case directly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (c)\n",
    "It is possible that the Fitness manifold has a barrier between position $(\\mathbf{x},F(\\mathbf{x}))$ and $(\\mathbf{y},F(\\mathbf{y}))$, i.e, for any path connecting the two points there is a region $B$ s,t, for any point $(\\mathbf{p},F(\\mathbf{p}))\\in B$, $F(\\mathbf{p})$ is very low. Therefore for any species try to move to ${\\mathbf{y}}$ from ${\\mathbf{x}}$, the low fitness score at the barrier may cause the extinction of this species during the transition.\n",
    "\n",
    "In the context of biology, it can be interperated as some part of the gene may simultatneously generate many optimal features and a risk of some disease of the species. Therefore, to get rid of the disease may first cause the loss of many good features on which the species heavily relies. Then, before the species regenerate the optimal features it need to survive, it could already be distinct in the natural selection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
